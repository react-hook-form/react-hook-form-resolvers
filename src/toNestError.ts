import {
  set,
  get,
  FieldError,
  FieldErrors,
  Field,
  InternalFieldName,
} from 'react-hook-form';

export const toNestError = <TFieldValues>(
  errors: Record<string, FieldError>,
  fields: Record<InternalFieldName, Field['_f']>,
): FieldErrors<TFieldValues> => {
  const fieldErrors: FieldErrors<TFieldValues> = {};
  for (const path in errors) {
    const field = get(fields, path) as Field['_f'] | undefined;
    set(
      fieldErrors,
      path,
      Object.assign(errors[path], { ref: field && field.ref }),
    );
  }

  return fieldErrors;
};
