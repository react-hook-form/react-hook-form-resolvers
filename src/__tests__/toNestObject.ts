import { Field, FieldError, InternalFieldName } from 'react-hook-form';
import { toNestError } from '../toNestError';

test('transforms flat object to nested object', () => {
  const flatObject: Record<string, FieldError> = {
    name: { type: 'st', message: 'first message' },
    'test.0.name': { type: 'nd', message: 'second message' },
    'n.test': { type: 'rd', message: 'third message' },
  };

  const fields = ({
    name: {
      ref: 'nameRef',
    },
    n: {
      test: {
        ref: 'testRef',
      },
    },
    unused: {
      ref: 'unusedRef',
    },
  } as any) as Record<InternalFieldName, Field['_f']>;

  expect(toNestError(flatObject, fields)).toMatchInlineSnapshot(`
    Object {
      "n": Object {
        "test": Object {
          "message": "third message",
          "ref": "testRef",
          "type": "rd",
        },
      },
      "name": Object {
        "message": "first message",
        "ref": "nameRef",
        "type": "st",
      },
      "test": Array [
        Object {
          "name": Object {
            "message": "second message",
            "ref": undefined,
            "type": "nd",
          },
        },
      ],
    }
  `);
});
