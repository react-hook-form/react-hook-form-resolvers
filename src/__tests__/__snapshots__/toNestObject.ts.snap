// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms flat object to nested object 1`] = `
Object {
  "name": Object {
    "message": "first message",
    "ref": Object {
      "reportValidity": [MockFunction],
      "setCustomValidity": [MockFunction],
    },
    "type": "st",
  },
  "test": Array [
    Object {
      "name": Object {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;

exports[`transforms flat object to nested object and shouldUseNativeValidation: true 1`] = `
Object {
  "name": Object {
    "message": "first message",
    "ref": Object {
      "reportValidity": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "setCustomValidity": [MockFunction] {
        "calls": Array [
          Array [
            "first message",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    },
    "type": "st",
  },
  "test": Array [
    Object {
      "name": Object {
        "message": "second message",
        "ref": undefined,
        "type": "nd",
      },
    },
  ],
}
`;
